impls/src/adapters/qr.rs

```diff
@@ -0,0 +1,210 @@
+		let method: CompressionFormat = match QR_VERSION {
+			// Version 0 had no compression method so it doesn't go here.
+			1 => CompressionFormat::Gzip,    // Latest version of emoji is version 1
+			2 => CompressionFormat::Zlib,    // just for example
+			_ => CompressionFormat::Deflate, // just for example
+		};
+
+		QrHeader {
+			algo: method,
+			version: QR_VERSION,
+		}
+	}
+
+	/// Returns a default value to modify without having to manually create a Header
+	fn new(ver: u8) -> QrHeader {
+		let method: CompressionFormat = match ver {
+			// Version 0 had no compression method so it doesn't go here.
+			1 => CompressionFormat::Gzip,    // Latest version of emoji is version 1
+			2 => CompressionFormat::Zlib,    // just for example
+			_ => CompressionFormat::Deflate, // just for example
+		};
+
+		QrHeader {
+			algo: method,
+			version: ver,
+		}
+	}
+}
+
+#[derive(Clone)]
+pub struct QrToSlate(pub PathBuf);
+
+/// This function will receive a skateboard as a `data` and save the QR code in image format based on `path_save`
+fn save2qr(data: &str, path_save: &str, receive_op: bool) -> Result<(), Error> {
+	// Header that will define the compression algorithm
+	let header = QrHeader::default();
+
+	// Compressing the data
+	let compressed_data = compress(data.as_bytes(), header.algo);
+
+	// Adding the QR code version at the beginning of the vector
+	let mut compressed_data_header: Vec<u8> = vec![QR_VERSION as u8];
+
+	// Adding the compressed data
+	compressed_data_header.extend(compressed_data);
+
+	// The response file is larger than send, so we limit to different scales
+	let (scalar, num_scalar) = if receive_op {
+		(1.0, 2)
+	} else {
+		(50.0 / 89.0, 1)
+	};
+
+	let num_out = data.matches("commit").count();
+	let num_ele: f32 = compressed_data_header.len() as f32;
+
+	// If have more than 4 commits => use more than 2 outputs to generate, because of that the QR can't be generated
+	// Also, the limit to QR is 2335 elements, so the response file is almost 1.78 = 89/50 biggest than the Send slate_json.
+	// So we limit the message size to be sent by 50 / 89 = length / (89/50)
+	if num_out > 3 * num_scalar || num_ele > LIMIT_QR_BIN * scalar {
+		error!("DataTooLong to generate the QR code! Try to perform the transaction by another method. The size of compressed Slate is: {:?}. The number of Outputs used is: {:?}",
+		num_ele,
+		num_out);
+
+		return Err(ErrorKind::DataTooLongQRError.into());
+	}
+
+	// Encode the data into bits
+	let code = QrCode::new(compressed_data_header).unwrap();
+
+	// Generate the image
+	let img = code.render::<Luma<u8>>().build();
+
+	// Save the image
+	img.save(path_save).unwrap();
+
+	Ok(())
+}
+
+/// This function will read an image and get all the QR code written and will transcribe it into a binary vector and at the end it will return the slate_json
+fn read_qr(path_read: &PathBuf) -> String {
+	// Open the image from disk
+	let img = image::open(path_read).expect("failed to open image");
```

Comment:
Can you improve this error message to also print the error itself from the image library?